File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/barrel/MilkingDisplay.java
package wraith.fabricaeexnihilo.compatibility.rei.barrel;

import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import me.shedaniel.rei.api.common.util.EntryIngredients;
import net.minecraft.item.ItemConvertible;
import net.minecraft.item.SpawnEggItem;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;
import wraith.fabricaeexnihilo.recipe.barrel.MilkingRecipe;

import java.util.List;
import java.util.Optional;
import java.util.stream.Stream;

@SuppressWarnings("UnstableApiUsage")
public final class MilkingDisplay implements Display {

    public final long amount;
    public final EntryIngredient entity;
    public final EntryIngredient result;
    private final Identifier id;

    public MilkingDisplay(MilkingRecipe recipe) {
        this.entity = EntryIngredients.ofItems(Stream.<ItemConvertible>ofNullable(SpawnEggItem.forEntity(recipe.getEntity())).toList());
        this.amount = recipe.getAmount();
        this.result = EntryIngredients.of(recipe.getFluid().getFluid(), amount);
        this.id = recipe.getId();
    }

    @Override
    public CategoryIdentifier<?> getCategoryIdentifier() {
        return PluginEntry.MILKING;
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return List.of(entity);
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.of(result);
    }

    @Override
    public Optional<Identifier> getDisplayLocation() {
        return Optional.of(id);
    }

}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/barrel/MilkingDisplay.kt
package exnihilofabrico.compatibility.rei.barrel

import exnihilofabrico.api.recipes.barrel.MilkingRecipe
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.modules.ModBlocks
import exnihilofabrico.util.asREIEntry
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeDisplay

class MilkingDisplay(val recipe: MilkingRecipe): RecipeDisplay {
    override fun getRecipeCategory() = PluginEntry.MILKING

    override fun getOutputEntries() = listOf(recipe.result.rawFluid?.bucketItem?.asREIEntry() ?: EntryStack.empty())
    override fun getInputEntries() =
        listOf(recipe.entity.asREIEntries(), ModBlocks.BARRELS.values.map { it.asREIEntry() })
}